metaflow/client/cache/cache_server.py

```diff
@@ -0,0 +1,307 @@
+import os
+import io
+import sys
+import json
+import fcntl
+import hashlib
+from itertools import chain
+from datetime import datetime
+from subprocess import Popen, PIPE
+from collections import deque, OrderedDict
+
+import click
+from metaflow.procpoll import make_poll
+
+from .cache_action import CacheAction,\
+                          LO_PRIO,\
+                          HI_PRIO,\
+                          import_action_class
+
+from .cache_store import CacheStore,\
+                         key_filename,\
+                         is_safely_readable
+
+class CacheServerException(Exception):
+    pass
+
+def server_request(op,
+                   action=None,
+                   prio=None,
+                   keys=None,
+                   stream_key=None,
+                   message=None,
+                   disposable_keys=None,
+                   idempotency_token=None):
+
+    if idempotency_token is None:
+        fields = [op]
+        if action:
+            fields.append(action)
+        if keys:
+            fields.extend(sorted(keys))
+        if stream_key:
+            fields.append(stream_key)
+        token = hashlib.sha1('|'.join(fields).encode('utf-8')).hexdigest()
+    else:
+        token = idempotency_token
+
+    return {
+        'op': op,
+        'action': action,
+        'priority': prio,
+        'keys': keys,
+        'stream_key': stream_key,
+        'message': message,
+        'idempotency_token': token,
+        'disposable_keys': disposable_keys
+    }
+
+def echo(msg):
+    now = datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%S.%f')
+    sys.stderr.write('CACHE [%s] %s\n' % (now, msg))
```

Comment:
is this a deliberate choice to always write output to stderr? f.ex. the CacheStore uses this to push out initialization information, so stderr seems like an odd place.

On a separate note: when running the async cache client, any of the messages produced by the subprocess echoes are not visible on-time, but rather only after an exception is encountered. For example the cache store initialization messages are only printed after a worker fails with an action. This might either be due to the use of stderr, or some asyncio shenanigans.